### PR TITLE
Provide complete refactoring of openDKIM script

### PR DESCRIPTION
**Pull request by aendeavor**
# Description

Complete refactoring of `target/bin/generate-dkim-config` and outsourcing of tests.

## What's left to do?

Many of the currently present tests need a rewrite. This is indicated in the `open_dkim.bats` file with `# TODO`. @wernerfred or @casperklein I have already spent more time than I have on this. I would need your help here.

**Note**: The `Makefile`'s `test` target is currently only testing the openDKIM config for speed. This will of course be reverted before merging.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or ENVIRONMENT.md or the Wiki)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -120,9 +120,11 @@ If you are using a LDAP setup the setup looks a bit different as you do not add
 ``` BASH
 docker-compose up -d mail
 
-./setup.sh config dkim <key-size> <domain.tld>[,<domain2.tld>]
+./setup.sh config dkim [size <size>] [selector <selector>] [domains '<domain.tld>[,<domain2.tld>]']
 ```
 
+If you want to see detailed usage information, run `./setup.sh dkim help`.
```
Comment (aendeavor) [inline README.md:126]: Here is a `config` after `setup.sh` and before `dkim` missing

File: target/bin/generate-dkim-config [hunk 2]
```diff
@@ -1,11 +1,88 @@
...
-KEYSIZE=${1:-4096}
-# optional domain names
-DOMAINS=${2:-}
+function __usage
+{
+  cat << END
+Generate DKIM Configuration
+
+SYNOPSIS
+    ./setup.sh dkim [OPTIONS...]
```
Comment (aendeavor) [inline target/bin/generate-dkim-config:13]: Here is a `config` missing as well

File: Makefile [hunk 3]
```diff
@@ -42,7 +44,8 @@ generate-accounts:
 	@ echo "           # this is also a test comment, :O" >> test/config/postfix-accounts.cf
 
 tests:
-	@ NAME=$(NAME) ./test/bats/bin/bats test/*.bats
+# @ NAME=$(NAME) ./test/bats/bin/bats test/*.bats
+	@ NAME=$(NAME) ./test/bats/bin/bats test/open_dkim.bats
```
Comment (wernerfred) [inline Makefile:48]: You did this to shorten up the CI it think?

**Comment by wernerfred** [comment]
Pretty busy today. I'll take a look at it as soon as I have a larger time frame available.
Last time i hotfixed the bad tests (told ya that calling them bad style is way no nice) i looked for alternatives and discovered that opendkim already provides a checkkey option. Sadly it relies on DNS i think which will not work (at least without changes). But I will see if I find a solution. 
If you came up with an idea on how to check the keys in a "correct" way feel free to drop it here.

File: Makefile (hunk 3 above)
Comment (aendeavor) [reply Makefile:48]: Yes, see the PR description

**Comment by aendeavor** [comment]
> Pretty busy today. I'll take a look at it as soon as I have a larger time frame available.

No hurry. Some tests just require proper formatting (I already did a bit on this topic), some a bit refactoring (get rid of `wc -l` and stuff), and some require more attention. Take your time.

I'm learning for exams atm, therefore I don not have any more spare time.

File: Makefile (hunk 3 above)
Comment (wernerfred) [reply Makefile:48]: My bad ;)

**Comment by polarathene** [comment]
After my PRs are sorted, I have a related `docker-compose` setup I plan to tidy up and publish on github as a reference setup. It has custom DNS service via container for all containers involved to use and another container that can update DNS records via a text template.

Might be appropriate for what you want to do, as I get the time, I'll be extending my reference to cover mail server DNS records myself. If you'd like to include for tests let me know :)

**Comment by aendeavor** [comment]
@wernerfred I would like to propose an idea: I would like to focus on the unit tests later in a separate PR because of two reasons:

1. I would like to see this PR merged :)
2. I don't want to pressure you into writing tests up. Currently tests are passing fine and I already separated them into their own file, which makes working with them in the future a lot easier.

If you agree, you can just go ahead and merge this or tell me and I will do it :)

**Comment by wernerfred** [comment]
Seems logic to me. I will agree as we will **definitely** need to refactor at least half of the test. Doing this in a separate step together with all others might safe more time in the end as doing it step by step and need to get into it every time again. But we should focus on this in the next time. 
This will be a massive PR on which multiple people can work for quite a while. Shouldn't affect the development of the project at all only rebases from time to time will be necessary.

**Review by wernerfred — APPROVED** [review]

**Comment by aendeavor** [comment]
> This will be a massive PR on which multiple people can work for quite a while. Shouldn't affect the development of the project at all only rebases from time to time will be necessary.

Completely agree.

File: target/bin/open-dkim [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        exit 2
+      fi
+      ;;
+
+    domain )
+      if [[ -n ${2+'set'} ]]
+      then
+        DOMAINS="${2}"
+        break
+        break
```
Comment (wernerfred) [inline target/bin/open-dkim:89]: Why are you using `break` here instead of `shift` ?

Comment (aendeavor) [reply target/bin/open-dkim:89]: Because I messed up. Can you correct this for me? Currently short on time...

Comment (wernerfred) [reply target/bin/open-dkim:89]: https://github.com/docker-mailserver/docker-mailserver/commit/6e6b5be1ee34fcce45448c1710f7d6a559bd1e54#diff-8ec5e15bee29bcdefebd50369ee7e19ad72d8b6e8c2d9475e3c4e8a463985a85

Comment (wernerfred) [reply target/bin/open-dkim:89]: https://github.com/docker-mailserver/docker-mailserver/runs/1929465064?check_suite_focus=true#step:6:325

@aendeavor any idea?:

```
# -- line differs --
# index    : 3
# expected :     open-dkim - configure DomainKeys Identified Mail (DKIM)
# actual   :     open-dkim - configure DomainKeys Identified Mail (DKIM)
# --
# 
```

Comment (wernerfred) [reply target/bin/open-dkim:89]: Actually your change (https://github.com/docker-mailserver/docker-mailserver/commit/d221c585c2adf9e08da72c440d21d0eb1b96bb00#diff-9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a089) never went through CI as it was no PR and therefore tests didn't run only the linting.

Comment (wernerfred) [reply target/bin/open-dkim:89]: Might the output have problems with colors?

Comment (wernerfred) [reply target/bin/open-dkim:89]: Tested it locally with `\e[0m` but no difference. So the problem might be something else. But I don't get it. The output is exactly the same.

Comment (wernerfred) [reply target/bin/open-dkim:89]: @docker-mailserver/maintainers any idea?

Comment (wernerfred) [reply target/bin/open-dkim:89]: So I added --partial to the assert line and the test went through locally, will push this as a hotfix but I want to know which "invisible" char breaks the check

Comment (aendeavor) [reply target/bin/open-dkim:89]: This is absolutely weird... For `setup.sh` it seems to work just fine...

**Pull request merged**